



<!DOCTYPE html>
<html lang="en" class="">
  <head prefix="og: http://ogp.me/ns# fb: http://ogp.me/ns/fb# object: http://ogp.me/ns/object# article: http://ogp.me/ns/article# profile: http://ogp.me/ns/profile#">
    <meta charset='utf-8'>

    <link crossorigin="anonymous" href="https://assets-cdn.github.com/assets/frameworks-94b6f3ff54dcc728ea8e0bb66a51433d9ac8f238ae6fff13829d767fcd15c261.css" media="all" rel="stylesheet" />
    <link crossorigin="anonymous" href="https://assets-cdn.github.com/assets/github-d8ae058d31b9f4426c2f16e381380775b7fe703d6471e7fb94572ce36b1b5b04.css" media="all" rel="stylesheet" />
    
    
    
    

    <link as="script" href="https://assets-cdn.github.com/assets/frameworks-4e4bd0cf3c442c849b926e75ec4e55233bc74e65771dc70d93aa133b32705488.js" rel="preload" />
    <link as="script" href="https://assets-cdn.github.com/assets/github-97fe248a925294b2a437a86d80688ff0684f57be07085dda8e893ff5d9a5338b.js" rel="preload" />

    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta http-equiv="Content-Language" content="en">
    <meta name="viewport" content="width=1020">
    
    
    <title>WegrzynLab/readme.md at master · SamGinzburg/WegrzynLab</title>
    <link rel="search" type="application/opensearchdescription+xml" href="/opensearch.xml" title="GitHub">
    <link rel="fluid-icon" href="https://github.com/fluidicon.png" title="GitHub">
    <link rel="apple-touch-icon" href="/apple-touch-icon.png">
    <link rel="apple-touch-icon" sizes="57x57" href="/apple-touch-icon-57x57.png">
    <link rel="apple-touch-icon" sizes="60x60" href="/apple-touch-icon-60x60.png">
    <link rel="apple-touch-icon" sizes="72x72" href="/apple-touch-icon-72x72.png">
    <link rel="apple-touch-icon" sizes="76x76" href="/apple-touch-icon-76x76.png">
    <link rel="apple-touch-icon" sizes="114x114" href="/apple-touch-icon-114x114.png">
    <link rel="apple-touch-icon" sizes="120x120" href="/apple-touch-icon-120x120.png">
    <link rel="apple-touch-icon" sizes="144x144" href="/apple-touch-icon-144x144.png">
    <link rel="apple-touch-icon" sizes="152x152" href="/apple-touch-icon-152x152.png">
    <link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon-180x180.png">
    <meta property="fb:app_id" content="1401488693436528">

      <meta content="https://avatars2.githubusercontent.com/u/9014527?v=3&amp;s=400" name="twitter:image:src" /><meta content="@github" name="twitter:site" /><meta content="summary" name="twitter:card" /><meta content="SamGinzburg/WegrzynLab" name="twitter:title" /><meta content="Contribute to WegrzynLab development by creating an account on GitHub." name="twitter:description" />
      <meta content="https://avatars2.githubusercontent.com/u/9014527?v=3&amp;s=400" property="og:image" /><meta content="GitHub" property="og:site_name" /><meta content="object" property="og:type" /><meta content="SamGinzburg/WegrzynLab" property="og:title" /><meta content="https://github.com/SamGinzburg/WegrzynLab" property="og:url" /><meta content="Contribute to WegrzynLab development by creating an account on GitHub." property="og:description" />
      <meta name="browser-stats-url" content="https://api.github.com/_private/browser/stats">
    <meta name="browser-errors-url" content="https://api.github.com/_private/browser/errors">
    <link rel="assets" href="https://assets-cdn.github.com/">
    <link rel="web-socket" href="wss://live.github.com/_sockets/MTEwMzM3NzU6MWEyOGQxMTVhYjE3NjQ4N2YwNzQ3OTY5N2VlMGM4YTQ6NWU4OTliOTQ2ZGYzNTAwOTUyYjE2OGM0NjM5NGE2Yzc5ZGY4ODYxMzBlNmE4MzdhNzMxNDM0MjUzYTZlZmI4Mg==--6f10e5529fbaeb7d38ec61bd2cfc545b576fdf33">
    <meta name="pjax-timeout" content="1000">
    <link rel="sudo-modal" href="/sessions/sudo_modal">

    <meta name="msapplication-TileImage" content="/windows-tile.png">
    <meta name="msapplication-TileColor" content="#ffffff">
    <meta name="selected-link" value="repo_source" data-pjax-transient>

    <meta name="google-site-verification" content="KT5gs8h0wvaagLKAVWq8bbeNwnZZK1r1XQysX3xurLU">
<meta name="google-site-verification" content="ZzhVyEFwb7w3e0-uOTltm8Jsck2F5StVihD0exw2fsA">
    <meta name="google-analytics" content="UA-3769691-2">

<meta content="collector.githubapp.com" name="octolytics-host" /><meta content="github" name="octolytics-app-id" /><meta content="896358C4:7CE0:22035872:56FC4C1C" name="octolytics-dimension-request_id" /><meta content="11033775" name="octolytics-actor-id" /><meta content="rop14009" name="octolytics-actor-login" /><meta content="68658d93d79cacff8ba6ac462a54edc0012d7623515464e01dbf73574f0c0759" name="octolytics-actor-hash" />
<meta content="/&lt;user-name&gt;/&lt;repo-name&gt;/blob/show" data-pjax-transient="true" name="analytics-location" />



  <meta class="js-ga-set" name="dimension1" content="Logged In">



        <meta name="hostname" content="github.com">
    <meta name="user-login" content="rop14009">

        <meta name="expected-hostname" content="github.com">
      <meta name="js-proxy-site-detection-payload" content="MmNmMTQ3MjBmMTQ3ZTM0ZTExM2UwYzA2NmE5ZmRiNjVkZDRkOTE1NTFjN2FhYTBhNTEyN2IxNDNkNWVhN2I2M3x7InJlbW90ZV9hZGRyZXNzIjoiMTM3Ljk5Ljg4LjE5NiIsInJlcXVlc3RfaWQiOiI4OTYzNThDNDo3Q0UwOjIyMDM1ODcyOjU2RkM0QzFDIiwidGltZXN0YW1wIjoxNDU5Mzc1MTMyfQ==">

      <link rel="mask-icon" href="https://assets-cdn.github.com/pinned-octocat.svg" color="#4078c0">
      <link rel="icon" type="image/x-icon" href="https://assets-cdn.github.com/favicon.ico">

    <meta content="78eeb736452018c702945640a1350d3a72d51eb0" name="form-nonce" />

    <meta http-equiv="x-pjax-version" content="37783c876cec67a6eb2599a31e2203bd">
    

      
  <meta name="description" content="Contribute to WegrzynLab development by creating an account on GitHub.">
  <meta name="go-import" content="github.com/SamGinzburg/WegrzynLab git https://github.com/SamGinzburg/WegrzynLab.git">

  <meta content="9014527" name="octolytics-dimension-user_id" /><meta content="SamGinzburg" name="octolytics-dimension-user_login" /><meta content="24800828" name="octolytics-dimension-repository_id" /><meta content="SamGinzburg/WegrzynLab" name="octolytics-dimension-repository_nwo" /><meta content="true" name="octolytics-dimension-repository_public" /><meta content="true" name="octolytics-dimension-repository_is_fork" /><meta content="24800503" name="octolytics-dimension-repository_parent_id" /><meta content="UConnBBC/WegrzynLab" name="octolytics-dimension-repository_parent_nwo" /><meta content="21360428" name="octolytics-dimension-repository_network_root_id" /><meta content="jcpickett1/WegrzynLab" name="octolytics-dimension-repository_network_root_nwo" />
  <link href="https://github.com/SamGinzburg/WegrzynLab/commits/master.atom" rel="alternate" title="Recent Commits to WegrzynLab:master" type="application/atom+xml">


      <link rel="canonical" href="https://github.com/SamGinzburg/WegrzynLab/blob/master/enTAP/readme.md" data-pjax-transient>
  </head>


  <body class="logged-in   env-production linux vis-public fork page-blob">
    <div id="js-pjax-loader-bar" class="pjax-loader-bar"></div>
    <a href="#start-of-content" tabindex="1" class="accessibility-aid js-skip-to-content">Skip to content</a>

    
    
    



        <div class="header header-logged-in true" role="banner">
  <div class="container clearfix">

    <a class="header-logo-invertocat" href="https://github.com/" data-hotkey="g d" aria-label="Homepage" data-ga-click="Header, go to dashboard, icon:logo">
  <svg aria-hidden="true" class="octicon octicon-mark-github" height="28" version="1.1" viewBox="0 0 16 16" width="28"><path d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59 0.4 0.07 0.55-0.17 0.55-0.38 0-0.19-0.01-0.82-0.01-1.49-2.01 0.37-2.53-0.49-2.69-0.94-0.09-0.23-0.48-0.94-0.82-1.13-0.28-0.15-0.68-0.52-0.01-0.53 0.63-0.01 1.08 0.58 1.23 0.82 0.72 1.21 1.87 0.87 2.33 0.66 0.07-0.52 0.28-0.87 0.51-1.07-1.78-0.2-3.64-0.89-3.64-3.95 0-0.87 0.31-1.59 0.82-2.15-0.08-0.2-0.36-1.02 0.08-2.12 0 0 0.67-0.21 2.2 0.82 0.64-0.18 1.32-0.27 2-0.27 0.68 0 1.36 0.09 2 0.27 1.53-1.04 2.2-0.82 2.2-0.82 0.44 1.1 0.16 1.92 0.08 2.12 0.51 0.56 0.82 1.27 0.82 2.15 0 3.07-1.87 3.75-3.65 3.95 0.29 0.25 0.54 0.73 0.54 1.48 0 1.07-0.01 1.93-0.01 2.2 0 0.21 0.15 0.46 0.55 0.38C13.71 14.53 16 11.53 16 8 16 3.58 12.42 0 8 0z"></path></svg>
</a>


        <div class="header-search scoped-search site-scoped-search js-site-search" role="search">
  <!-- </textarea> --><!-- '"` --><form accept-charset="UTF-8" action="/SamGinzburg/WegrzynLab/search" class="js-site-search-form" data-scoped-search-url="/SamGinzburg/WegrzynLab/search" data-unscoped-search-url="/search" method="get"><div style="margin:0;padding:0;display:inline"><input name="utf8" type="hidden" value="&#x2713;" /></div>
    <label class="form-control header-search-wrapper js-chromeless-input-container">
      <div class="header-search-scope">This repository</div>
      <input type="text"
        class="form-control header-search-input js-site-search-focus js-site-search-field is-clearable"
        data-hotkey="s"
        name="q"
        placeholder="Search"
        aria-label="Search this repository"
        data-unscoped-placeholder="Search GitHub"
        data-scoped-placeholder="Search"
        tabindex="1"
        autocapitalize="off">
    </label>
</form></div>


      <ul class="header-nav left" role="navigation">
        <li class="header-nav-item">
          <a href="/pulls" class="js-selected-navigation-item header-nav-link" data-ga-click="Header, click, Nav menu - item:pulls context:user" data-hotkey="g p" data-selected-links="/pulls /pulls/assigned /pulls/mentioned /pulls">
            Pull requests
</a>        </li>
        <li class="header-nav-item">
          <a href="/issues" class="js-selected-navigation-item header-nav-link" data-ga-click="Header, click, Nav menu - item:issues context:user" data-hotkey="g i" data-selected-links="/issues /issues/assigned /issues/mentioned /issues">
            Issues
</a>        </li>
          <li class="header-nav-item">
            <a class="header-nav-link" href="https://gist.github.com/" data-ga-click="Header, go to gist, text:gist">Gist</a>
          </li>
      </ul>

    
<ul class="header-nav user-nav right" id="user-links">
  <li class="header-nav-item">
    
    <a href="/notifications" aria-label="You have no unread notifications" class="header-nav-link notification-indicator tooltipped tooltipped-s js-socket-channel js-notification-indicator" data-channel="notification-changed-v2:11033775" data-ga-click="Header, go to notifications, icon:read" data-hotkey="g n">
        <span class="mail-status "></span>
        <svg aria-hidden="true" class="octicon octicon-bell" height="16" version="1.1" viewBox="0 0 14 16" width="14"><path d="M14 12v1H0v-1l0.73-0.58c0.77-0.77 0.81-2.55 1.19-4.42 0.77-3.77 4.08-5 4.08-5 0-0.55 0.45-1 1-1s1 0.45 1 1c0 0 3.39 1.23 4.16 5 0.38 1.88 0.42 3.66 1.19 4.42l0.66 0.58z m-7 4c1.11 0 2-0.89 2-2H5c0 1.11 0.89 2 2 2z"></path></svg>
</a>
  </li>

  <li class="header-nav-item dropdown js-menu-container">
    <a class="header-nav-link tooltipped tooltipped-s js-menu-target" href="/new"
       aria-label="Create new…"
       data-ga-click="Header, create new, icon:add">
      <svg aria-hidden="true" class="octicon octicon-plus left" height="16" version="1.1" viewBox="0 0 12 16" width="12"><path d="M12 9H7v5H5V9H0V7h5V2h2v5h5v2z"></path></svg>
      <span class="dropdown-caret"></span>
    </a>

    <div class="dropdown-menu-content js-menu-content">
      <ul class="dropdown-menu dropdown-menu-sw">
        
<a class="dropdown-item" href="/new" data-ga-click="Header, create new repository">
  New repository
</a>


  <a class="dropdown-item" href="/organizations/new" data-ga-click="Header, create new organization">
    New organization
  </a>




      </ul>
    </div>
  </li>

  <li class="header-nav-item dropdown js-menu-container">
    <a class="header-nav-link name tooltipped tooltipped-sw js-menu-target" href="/rop14009"
       aria-label="View profile and more"
       data-ga-click="Header, show menu, icon:avatar">
      <img alt="@rop14009" class="avatar" height="20" src="https://avatars1.githubusercontent.com/u/11033775?v=3&amp;s=40" width="20" />
      <span class="dropdown-caret"></span>
    </a>

    <div class="dropdown-menu-content js-menu-content">
      <div class="dropdown-menu  dropdown-menu-sw">
        <div class=" dropdown-header header-nav-current-user css-truncate">
            Signed in as <strong class="css-truncate-target">rop14009</strong>

        </div>


        <div class="dropdown-divider"></div>

          <a class="dropdown-item" href="/rop14009" data-ga-click="Header, go to profile, text:your profile">
            Your profile
          </a>
        <a class="dropdown-item" href="/stars" data-ga-click="Header, go to starred repos, text:your stars">
          Your stars
        </a>
          <a class="dropdown-item" href="/explore" data-ga-click="Header, go to explore, text:explore">
            Explore
          </a>
          <a class="dropdown-item" href="/integrations" data-ga-click="Header, go to integrations, text:integrations">
            Integrations
          </a>
        <a class="dropdown-item" href="https://help.github.com" data-ga-click="Header, go to help, text:help">
          Help
        </a>


          <div class="dropdown-divider"></div>

          <a class="dropdown-item" href="/settings/profile" data-ga-click="Header, go to settings, icon:settings">
            Settings
          </a>

          <!-- </textarea> --><!-- '"` --><form accept-charset="UTF-8" action="/logout" class="logout-form" data-form-nonce="78eeb736452018c702945640a1350d3a72d51eb0" method="post"><div style="margin:0;padding:0;display:inline"><input name="utf8" type="hidden" value="&#x2713;" /><input name="authenticity_token" type="hidden" value="kJEgvt+4orv0ow4c5QYluXc3zK2Bx6R3DtZQSrsSHlAd/ozFFX04JSiHqGFU/beGRR9sNmXjW7dN6HleXvMdxg==" /></div>
            <button class="dropdown-item dropdown-signout" data-ga-click="Header, sign out, icon:logout">
              Sign out
            </button>
</form>
      </div>
    </div>
  </li>
</ul>


    
  </div>
</div>


      


    <div id="start-of-content" class="accessibility-aid"></div>

      <div id="js-flash-container">
</div>


    <div role="main" class="main-content">
        <div itemscope itemtype="http://schema.org/SoftwareSourceCode">
    <div id="js-repo-pjax-container" data-pjax-container>
      
<div class="pagehead repohead instapaper_ignore readability-menu experiment-repo-nav">
  <div class="container repohead-details-container">

    

<ul class="pagehead-actions">

  <li>
        <!-- </textarea> --><!-- '"` --><form accept-charset="UTF-8" action="/notifications/subscribe" class="js-social-container" data-autosubmit="true" data-form-nonce="78eeb736452018c702945640a1350d3a72d51eb0" data-remote="true" method="post"><div style="margin:0;padding:0;display:inline"><input name="utf8" type="hidden" value="&#x2713;" /><input name="authenticity_token" type="hidden" value="xeS9o42H8OxSiJYg8GAjJHSo7KL87uMgyjKFpP+C2hkPrDsVb1Xi7OVhKd1RvmseZh/UIcZWPRR//t7Tdn5jCg==" /></div>      <input class="form-control" id="repository_id" name="repository_id" type="hidden" value="24800828" />

        <div class="select-menu js-menu-container js-select-menu">
          <a href="/SamGinzburg/WegrzynLab/subscription"
            class="btn btn-sm btn-with-count select-menu-button js-menu-target" role="button" tabindex="0" aria-haspopup="true"
            data-ga-click="Repository, click Watch settings, action:blob#show">
            <span class="js-select-button">
              <svg aria-hidden="true" class="octicon octicon-eye" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M8.06 2C3 2 0 8 0 8s3 6 8.06 6c4.94 0 7.94-6 7.94-6S13 2 8.06 2z m-0.06 10c-2.2 0-4-1.78-4-4 0-2.2 1.8-4 4-4 2.22 0 4 1.8 4 4 0 2.22-1.78 4-4 4z m2-4c0 1.11-0.89 2-2 2s-2-0.89-2-2 0.89-2 2-2 2 0.89 2 2z"></path></svg>
              Watch
            </span>
          </a>
          <a class="social-count js-social-count" href="/SamGinzburg/WegrzynLab/watchers">
            3
          </a>

        <div class="select-menu-modal-holder">
          <div class="select-menu-modal subscription-menu-modal js-menu-content" aria-hidden="true">
            <div class="select-menu-header js-navigation-enable" tabindex="-1">
              <svg aria-label="Close" class="octicon octicon-x js-menu-close" height="16" role="img" version="1.1" viewBox="0 0 12 16" width="12"><path d="M7.48 8l3.75 3.75-1.48 1.48-3.75-3.75-3.75 3.75-1.48-1.48 3.75-3.75L0.77 4.25l1.48-1.48 3.75 3.75 3.75-3.75 1.48 1.48-3.75 3.75z"></path></svg>
              <span class="select-menu-title">Notifications</span>
            </div>

              <div class="select-menu-list js-navigation-container" role="menu">

                <div class="select-menu-item js-navigation-item selected" role="menuitem" tabindex="0">
                  <svg aria-hidden="true" class="octicon octicon-check select-menu-item-icon" height="16" version="1.1" viewBox="0 0 12 16" width="12"><path d="M12 5L4 13 0 9l1.5-1.5 2.5 2.5 6.5-6.5 1.5 1.5z"></path></svg>
                  <div class="select-menu-item-text">
                    <input checked="checked" id="do_included" name="do" type="radio" value="included" />
                    <span class="select-menu-item-heading">Not watching</span>
                    <span class="description">Be notified when participating or @mentioned.</span>
                    <span class="js-select-button-text hidden-select-button-text">
                      <svg aria-hidden="true" class="octicon octicon-eye" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M8.06 2C3 2 0 8 0 8s3 6 8.06 6c4.94 0 7.94-6 7.94-6S13 2 8.06 2z m-0.06 10c-2.2 0-4-1.78-4-4 0-2.2 1.8-4 4-4 2.22 0 4 1.8 4 4 0 2.22-1.78 4-4 4z m2-4c0 1.11-0.89 2-2 2s-2-0.89-2-2 0.89-2 2-2 2 0.89 2 2z"></path></svg>
                      Watch
                    </span>
                  </div>
                </div>

                <div class="select-menu-item js-navigation-item " role="menuitem" tabindex="0">
                  <svg aria-hidden="true" class="octicon octicon-check select-menu-item-icon" height="16" version="1.1" viewBox="0 0 12 16" width="12"><path d="M12 5L4 13 0 9l1.5-1.5 2.5 2.5 6.5-6.5 1.5 1.5z"></path></svg>
                  <div class="select-menu-item-text">
                    <input id="do_subscribed" name="do" type="radio" value="subscribed" />
                    <span class="select-menu-item-heading">Watching</span>
                    <span class="description">Be notified of all conversations.</span>
                    <span class="js-select-button-text hidden-select-button-text">
                      <svg aria-hidden="true" class="octicon octicon-eye" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M8.06 2C3 2 0 8 0 8s3 6 8.06 6c4.94 0 7.94-6 7.94-6S13 2 8.06 2z m-0.06 10c-2.2 0-4-1.78-4-4 0-2.2 1.8-4 4-4 2.22 0 4 1.8 4 4 0 2.22-1.78 4-4 4z m2-4c0 1.11-0.89 2-2 2s-2-0.89-2-2 0.89-2 2-2 2 0.89 2 2z"></path></svg>
                      Unwatch
                    </span>
                  </div>
                </div>

                <div class="select-menu-item js-navigation-item " role="menuitem" tabindex="0">
                  <svg aria-hidden="true" class="octicon octicon-check select-menu-item-icon" height="16" version="1.1" viewBox="0 0 12 16" width="12"><path d="M12 5L4 13 0 9l1.5-1.5 2.5 2.5 6.5-6.5 1.5 1.5z"></path></svg>
                  <div class="select-menu-item-text">
                    <input id="do_ignore" name="do" type="radio" value="ignore" />
                    <span class="select-menu-item-heading">Ignoring</span>
                    <span class="description">Never be notified.</span>
                    <span class="js-select-button-text hidden-select-button-text">
                      <svg aria-hidden="true" class="octicon octicon-mute" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M8 2.81v10.38c0 0.67-0.81 1-1.28 0.53L3 10H1c-0.55 0-1-0.45-1-1V7c0-0.55 0.45-1 1-1h2l3.72-3.72c0.47-0.47 1.28-0.14 1.28 0.53z m7.53 3.22l-1.06-1.06-1.97 1.97-1.97-1.97-1.06 1.06 1.97 1.97-1.97 1.97 1.06 1.06 1.97-1.97 1.97 1.97 1.06-1.06-1.97-1.97 1.97-1.97z"></path></svg>
                      Stop ignoring
                    </span>
                  </div>
                </div>

              </div>

            </div>
          </div>
        </div>
</form>
  </li>

  <li>
    
  <div class="js-toggler-container js-social-container starring-container ">

    <!-- </textarea> --><!-- '"` --><form accept-charset="UTF-8" action="/SamGinzburg/WegrzynLab/unstar" class="js-toggler-form starred" data-form-nonce="78eeb736452018c702945640a1350d3a72d51eb0" data-remote="true" method="post"><div style="margin:0;padding:0;display:inline"><input name="utf8" type="hidden" value="&#x2713;" /><input name="authenticity_token" type="hidden" value="DbcbMP5muKv0pUniUVSngRw9tOreu0wAa9UgxjvD/2V5LTE7/K2UaievmkO583MrC45Szyveh+aTWtsxjmqcqA==" /></div>
      <button
        class="btn btn-sm btn-with-count js-toggler-target"
        aria-label="Unstar this repository" title="Unstar SamGinzburg/WegrzynLab"
        data-ga-click="Repository, click unstar button, action:blob#show; text:Unstar">
        <svg aria-hidden="true" class="octicon octicon-star" height="16" version="1.1" viewBox="0 0 14 16" width="14"><path d="M14 6l-4.9-0.64L7 1 4.9 5.36 0 6l3.6 3.26L2.67 14l4.33-2.33 4.33 2.33L10.4 9.26 14 6z"></path></svg>
        Unstar
      </button>
        <a class="social-count js-social-count" href="/SamGinzburg/WegrzynLab/stargazers">
          1
        </a>
</form>
    <!-- </textarea> --><!-- '"` --><form accept-charset="UTF-8" action="/SamGinzburg/WegrzynLab/star" class="js-toggler-form unstarred" data-form-nonce="78eeb736452018c702945640a1350d3a72d51eb0" data-remote="true" method="post"><div style="margin:0;padding:0;display:inline"><input name="utf8" type="hidden" value="&#x2713;" /><input name="authenticity_token" type="hidden" value="EWubPspCDvcAhDZSvq848xFBWdEX1l6cGc+5dReP0zbWPqjJB8q9CZ/7Tot+m8rq1aT6FDAWirKQiYdaWsHkaQ==" /></div>
      <button
        class="btn btn-sm btn-with-count js-toggler-target"
        aria-label="Star this repository" title="Star SamGinzburg/WegrzynLab"
        data-ga-click="Repository, click star button, action:blob#show; text:Star">
        <svg aria-hidden="true" class="octicon octicon-star" height="16" version="1.1" viewBox="0 0 14 16" width="14"><path d="M14 6l-4.9-0.64L7 1 4.9 5.36 0 6l3.6 3.26L2.67 14l4.33-2.33 4.33 2.33L10.4 9.26 14 6z"></path></svg>
        Star
      </button>
        <a class="social-count js-social-count" href="/SamGinzburg/WegrzynLab/stargazers">
          1
        </a>
</form>  </div>

  </li>

  <li>
          <!-- </textarea> --><!-- '"` --><form accept-charset="UTF-8" action="/SamGinzburg/WegrzynLab/fork" class="btn-with-count" data-form-nonce="78eeb736452018c702945640a1350d3a72d51eb0" method="post"><div style="margin:0;padding:0;display:inline"><input name="utf8" type="hidden" value="&#x2713;" /><input name="authenticity_token" type="hidden" value="4pNAYEtoyrmmKFOe1cQfX6qTOeleO8Ow4TJ52mSQe3Hvzm152/Qm6/e2TFife7rBI+FQk1mNB3BzeKn85/Nh8Q==" /></div>
            <button
                type="submit"
                class="btn btn-sm btn-with-count"
                data-ga-click="Repository, show fork modal, action:blob#show; text:Fork"
                title="Fork your own copy of SamGinzburg/WegrzynLab to your account"
                aria-label="Fork your own copy of SamGinzburg/WegrzynLab to your account">
              <svg aria-hidden="true" class="octicon octicon-repo-forked" height="16" version="1.1" viewBox="0 0 10 16" width="10"><path d="M8 1c-1.11 0-2 0.89-2 2 0 0.73 0.41 1.38 1 1.72v1.28L5 8 3 6v-1.28c0.59-0.34 1-0.98 1-1.72 0-1.11-0.89-2-2-2S0 1.89 0 3c0 0.73 0.41 1.38 1 1.72v1.78l3 3v1.78c-0.59 0.34-1 0.98-1 1.72 0 1.11 0.89 2 2 2s2-0.89 2-2c0-0.73-0.41-1.38-1-1.72V9.5l3-3V4.72c0.59-0.34 1-0.98 1-1.72 0-1.11-0.89-2-2-2zM2 4.2c-0.66 0-1.2-0.55-1.2-1.2s0.55-1.2 1.2-1.2 1.2 0.55 1.2 1.2-0.55 1.2-1.2 1.2z m3 10c-0.66 0-1.2-0.55-1.2-1.2s0.55-1.2 1.2-1.2 1.2 0.55 1.2 1.2-0.55 1.2-1.2 1.2z m3-10c-0.66 0-1.2-0.55-1.2-1.2s0.55-1.2 1.2-1.2 1.2 0.55 1.2 1.2-0.55 1.2-1.2 1.2z"></path></svg>
              Fork
            </button>
</form>
    <a href="/SamGinzburg/WegrzynLab/network" class="social-count">
      4
    </a>
  </li>
</ul>

    <h1 class="entry-title public ">
  <svg aria-hidden="true" class="octicon octicon-repo-forked" height="16" version="1.1" viewBox="0 0 10 16" width="10"><path d="M8 1c-1.11 0-2 0.89-2 2 0 0.73 0.41 1.38 1 1.72v1.28L5 8 3 6v-1.28c0.59-0.34 1-0.98 1-1.72 0-1.11-0.89-2-2-2S0 1.89 0 3c0 0.73 0.41 1.38 1 1.72v1.78l3 3v1.78c-0.59 0.34-1 0.98-1 1.72 0 1.11 0.89 2 2 2s2-0.89 2-2c0-0.73-0.41-1.38-1-1.72V9.5l3-3V4.72c0.59-0.34 1-0.98 1-1.72 0-1.11-0.89-2-2-2zM2 4.2c-0.66 0-1.2-0.55-1.2-1.2s0.55-1.2 1.2-1.2 1.2 0.55 1.2 1.2-0.55 1.2-1.2 1.2z m3 10c-0.66 0-1.2-0.55-1.2-1.2s0.55-1.2 1.2-1.2 1.2 0.55 1.2 1.2-0.55 1.2-1.2 1.2z m3-10c-0.66 0-1.2-0.55-1.2-1.2s0.55-1.2 1.2-1.2 1.2 0.55 1.2 1.2-0.55 1.2-1.2 1.2z"></path></svg>
  <span class="author" itemprop="author"><a href="/SamGinzburg" class="url fn" rel="author">SamGinzburg</a></span><!--
--><span class="path-divider">/</span><!--
--><strong itemprop="name"><a href="/SamGinzburg/WegrzynLab" data-pjax="#js-repo-pjax-container">WegrzynLab</a></strong>

    <span class="fork-flag">
      <span class="text">forked from <a href="/UConnBBC/WegrzynLab">UConnBBC/WegrzynLab</a></span>
    </span>
</h1>

  </div>
  <div class="container">
    
<nav class="reponav js-repo-nav js-sidenav-container-pjax"
     itemscope
     itemtype="http://schema.org/BreadcrumbList"
     role="navigation"
     data-pjax="#js-repo-pjax-container">

  <span itemscope itemtype="http://schema.org/ListItem" itemprop="itemListElement">
    <a href="/SamGinzburg/WegrzynLab" aria-selected="true" class="js-selected-navigation-item selected reponav-item" data-hotkey="g c" data-selected-links="repo_source repo_downloads repo_commits repo_releases repo_tags repo_branches /SamGinzburg/WegrzynLab" itemprop="url">
      <svg aria-hidden="true" class="octicon octicon-code" height="16" version="1.1" viewBox="0 0 14 16" width="14"><path d="M9.5 3l-1.5 1.5 3.5 3.5L8 11.5l1.5 1.5 4.5-5L9.5 3zM4.5 3L0 8l4.5 5 1.5-1.5L2.5 8l3.5-3.5L4.5 3z"></path></svg>
      <span itemprop="name">Code</span>
      <meta itemprop="position" content="1">
</a>  </span>


  <span itemscope itemtype="http://schema.org/ListItem" itemprop="itemListElement">
    <a href="/SamGinzburg/WegrzynLab/pulls" class="js-selected-navigation-item reponav-item" data-hotkey="g p" data-selected-links="repo_pulls /SamGinzburg/WegrzynLab/pulls" itemprop="url">
      <svg aria-hidden="true" class="octicon octicon-git-pull-request" height="16" version="1.1" viewBox="0 0 12 16" width="12"><path d="M11 11.28c0-1.73 0-6.28 0-6.28-0.03-0.78-0.34-1.47-0.94-2.06s-1.28-0.91-2.06-0.94c0 0-1.02 0-1 0V0L4 3l3 3V4h1c0.27 0.02 0.48 0.11 0.69 0.31s0.3 0.42 0.31 0.69v6.28c-0.59 0.34-1 0.98-1 1.72 0 1.11 0.89 2 2 2s2-0.89 2-2c0-0.73-0.41-1.38-1-1.72z m-1 2.92c-0.66 0-1.2-0.55-1.2-1.2s0.55-1.2 1.2-1.2 1.2 0.55 1.2 1.2-0.55 1.2-1.2 1.2zM4 3c0-1.11-0.89-2-2-2S0 1.89 0 3c0 0.73 0.41 1.38 1 1.72 0 1.55 0 5.56 0 6.56-0.59 0.34-1 0.98-1 1.72 0 1.11 0.89 2 2 2s2-0.89 2-2c0-0.73-0.41-1.38-1-1.72V4.72c0.59-0.34 1-0.98 1-1.72z m-0.8 10c0 0.66-0.55 1.2-1.2 1.2s-1.2-0.55-1.2-1.2 0.55-1.2 1.2-1.2 1.2 0.55 1.2 1.2z m-1.2-8.8c-0.66 0-1.2-0.55-1.2-1.2s0.55-1.2 1.2-1.2 1.2 0.55 1.2 1.2-0.55 1.2-1.2 1.2z"></path></svg>
      <span itemprop="name">Pull requests</span>
      <span class="counter">0</span>
      <meta itemprop="position" content="3">
</a>  </span>


  <a href="/SamGinzburg/WegrzynLab/pulse" class="js-selected-navigation-item reponav-item" data-selected-links="pulse /SamGinzburg/WegrzynLab/pulse">
    <svg aria-hidden="true" class="octicon octicon-pulse" height="16" version="1.1" viewBox="0 0 14 16" width="14"><path d="M11.5 8L8.8 5.4 6.6 8.5 5.5 1.6 2.38 8H0V10h3.6L4.5 8.2l0.9 5.4L9 8.5l1.6 1.5H14V8H11.5z"></path></svg>
    Pulse
</a>
  <a href="/SamGinzburg/WegrzynLab/graphs" class="js-selected-navigation-item reponav-item" data-selected-links="repo_graphs repo_contributors /SamGinzburg/WegrzynLab/graphs">
    <svg aria-hidden="true" class="octicon octicon-graph" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M16 14v1H0V0h1v14h15z m-11-1H3V8h2v5z m4 0H7V3h2v10z m4 0H11V6h2v7z"></path></svg>
    Graphs
</a>

</nav>

  </div>
</div>

<div class="container new-discussion-timeline experiment-repo-nav">
  <div class="repository-content">

    

<a href="/SamGinzburg/WegrzynLab/blob/a5f3cd7f180e3d87422abdd8a147581aec60f934/enTAP/readme.md" class="hidden js-permalink-shortcut" data-hotkey="y">Permalink</a>

<!-- blob contrib key: blob_contributors:v21:86c828379090ef7e5bfc5708fa8e2a18 -->

<div class="file-navigation js-zeroclipboard-container">
  
<div class="select-menu js-menu-container js-select-menu left">
  <button class="btn btn-sm select-menu-button js-menu-target css-truncate" data-hotkey="w"
    title="master"
    type="button" aria-label="Switch branches or tags" tabindex="0" aria-haspopup="true">
    <i>Branch:</i>
    <span class="js-select-button css-truncate-target">master</span>
  </button>

  <div class="select-menu-modal-holder js-menu-content js-navigation-container" data-pjax aria-hidden="true">

    <div class="select-menu-modal">
      <div class="select-menu-header">
        <svg aria-label="Close" class="octicon octicon-x js-menu-close" height="16" role="img" version="1.1" viewBox="0 0 12 16" width="12"><path d="M7.48 8l3.75 3.75-1.48 1.48-3.75-3.75-3.75 3.75-1.48-1.48 3.75-3.75L0.77 4.25l1.48-1.48 3.75 3.75 3.75-3.75 1.48 1.48-3.75 3.75z"></path></svg>
        <span class="select-menu-title">Switch branches/tags</span>
      </div>

      <div class="select-menu-filters">
        <div class="select-menu-text-filter">
          <input type="text" aria-label="Filter branches/tags" id="context-commitish-filter-field" class="form-control js-filterable-field js-navigation-enable" placeholder="Filter branches/tags">
        </div>
        <div class="select-menu-tabs">
          <ul>
            <li class="select-menu-tab">
              <a href="#" data-tab-filter="branches" data-filter-placeholder="Filter branches/tags" class="js-select-menu-tab" role="tab">Branches</a>
            </li>
            <li class="select-menu-tab">
              <a href="#" data-tab-filter="tags" data-filter-placeholder="Find a tag…" class="js-select-menu-tab" role="tab">Tags</a>
            </li>
          </ul>
        </div>
      </div>

      <div class="select-menu-list select-menu-tab-bucket js-select-menu-tab-bucket" data-tab-filter="branches" role="menu">

        <div data-filterable-for="context-commitish-filter-field" data-filterable-type="substring">


            <a class="select-menu-item js-navigation-item js-navigation-open "
               href="/SamGinzburg/WegrzynLab/blob/drafts/enTAP/readme.md"
               data-name="drafts"
               data-skip-pjax="true"
               rel="nofollow">
              <svg aria-hidden="true" class="octicon octicon-check select-menu-item-icon" height="16" version="1.1" viewBox="0 0 12 16" width="12"><path d="M12 5L4 13 0 9l1.5-1.5 2.5 2.5 6.5-6.5 1.5 1.5z"></path></svg>
              <span class="select-menu-item-text css-truncate-target js-select-menu-filter-text" title="drafts">
                drafts
              </span>
            </a>
            <a class="select-menu-item js-navigation-item js-navigation-open selected"
               href="/SamGinzburg/WegrzynLab/blob/master/enTAP/readme.md"
               data-name="master"
               data-skip-pjax="true"
               rel="nofollow">
              <svg aria-hidden="true" class="octicon octicon-check select-menu-item-icon" height="16" version="1.1" viewBox="0 0 12 16" width="12"><path d="M12 5L4 13 0 9l1.5-1.5 2.5 2.5 6.5-6.5 1.5 1.5z"></path></svg>
              <span class="select-menu-item-text css-truncate-target js-select-menu-filter-text" title="master">
                master
              </span>
            </a>
        </div>

          <div class="select-menu-no-results">Nothing to show</div>
      </div>

      <div class="select-menu-list select-menu-tab-bucket js-select-menu-tab-bucket" data-tab-filter="tags">
        <div data-filterable-for="context-commitish-filter-field" data-filterable-type="substring">


        </div>

        <div class="select-menu-no-results">Nothing to show</div>
      </div>

    </div>
  </div>
</div>

  <div class="btn-group right">
    <a href="/SamGinzburg/WegrzynLab/find/master"
          class="js-pjax-capture-input btn btn-sm"
          data-pjax
          data-hotkey="t">
      Find file
    </a>
    <button aria-label="Copy file path to clipboard" class="js-zeroclipboard btn btn-sm zeroclipboard-button tooltipped tooltipped-s" data-copied-hint="Copied!" type="button">Copy path</button>
  </div>
  <div class="breadcrumb js-zeroclipboard-target">
    <span class="repo-root js-repo-root"><span class="js-path-segment"><a href="/SamGinzburg/WegrzynLab"><span>WegrzynLab</span></a></span></span><span class="separator">/</span><span class="js-path-segment"><a href="/SamGinzburg/WegrzynLab/tree/master/enTAP"><span>enTAP</span></a></span><span class="separator">/</span><strong class="final-path">readme.md</strong>
  </div>
</div>

<include-fragment class="commit-tease" src="/SamGinzburg/WegrzynLab/contributors/master/enTAP/readme.md">
  <div>
    Fetching contributors&hellip;
  </div>

  <div class="commit-tease-contributors">
    <img alt="" class="loader-loading left" height="16" src="https://assets-cdn.github.com/images/spinners/octocat-spinner-32-EAF2F5.gif" width="16" />
    <span class="loader-error">Cannot retrieve contributors at this time</span>
  </div>
</include-fragment>
<div class="file">
  <div class="file-header">
  <div class="file-actions">

    <div class="btn-group">
      <a href="/SamGinzburg/WegrzynLab/raw/master/enTAP/readme.md" class="btn btn-sm " id="raw-url">Raw</a>
        <a href="/SamGinzburg/WegrzynLab/blame/master/enTAP/readme.md" class="btn btn-sm js-update-url-with-hash">Blame</a>
      <a href="/SamGinzburg/WegrzynLab/commits/master/enTAP/readme.md" class="btn btn-sm " rel="nofollow">History</a>
    </div>


        <!-- </textarea> --><!-- '"` --><form accept-charset="UTF-8" action="/SamGinzburg/WegrzynLab/edit/master/enTAP/readme.md" class="inline-form js-update-url-with-hash" data-form-nonce="78eeb736452018c702945640a1350d3a72d51eb0" method="post"><div style="margin:0;padding:0;display:inline"><input name="utf8" type="hidden" value="&#x2713;" /><input name="authenticity_token" type="hidden" value="ywvDmlztE/pENOOaRrDJGQThAaLA2EBhfQrtghGE/RpiCrvPaw8q9mDeXuWkLYdiTgth4VHnqTb7NMH6Z14T2g==" /></div>
          <button class="btn-octicon tooltipped tooltipped-nw" type="submit"
            aria-label="Fork this project and edit the file" data-hotkey="e" data-disable-with>
            <svg aria-hidden="true" class="octicon octicon-pencil" height="16" version="1.1" viewBox="0 0 14 16" width="14"><path d="M0 12v3h3l8-8-3-3L0 12z m3 2H1V12h1v1h1v1z m10.3-9.3l-1.3 1.3-3-3 1.3-1.3c0.39-0.39 1.02-0.39 1.41 0l1.59 1.59c0.39 0.39 0.39 1.02 0 1.41z"></path></svg>
          </button>
</form>        <!-- </textarea> --><!-- '"` --><form accept-charset="UTF-8" action="/SamGinzburg/WegrzynLab/delete/master/enTAP/readme.md" class="inline-form" data-form-nonce="78eeb736452018c702945640a1350d3a72d51eb0" method="post"><div style="margin:0;padding:0;display:inline"><input name="utf8" type="hidden" value="&#x2713;" /><input name="authenticity_token" type="hidden" value="7juoSK+VpS2JLdOW7PVnDSvEsxvoQDl+192Vfcq3AshiGrwIPD57GCtfq3TeZNWYArb408n1+015AI+r/Fn/Gg==" /></div>
          <button class="btn-octicon btn-octicon-danger tooltipped tooltipped-nw" type="submit"
            aria-label="Fork this project and delete the file" data-disable-with>
            <svg aria-hidden="true" class="octicon octicon-trashcan" height="16" version="1.1" viewBox="0 0 12 16" width="12"><path d="M10 2H8c0-0.55-0.45-1-1-1H4c-0.55 0-1 0.45-1 1H1c-0.55 0-1 0.45-1 1v1c0 0.55 0.45 1 1 1v9c0 0.55 0.45 1 1 1h7c0.55 0 1-0.45 1-1V5c0.55 0 1-0.45 1-1v-1c0-0.55-0.45-1-1-1z m-1 12H2V5h1v8h1V5h1v8h1V5h1v8h1V5h1v9z m1-10H1v-1h9v1z"></path></svg>
          </button>
</form>  </div>

  <div class="file-info">
      519 lines (376 sloc)
      <span class="file-info-divider"></span>
    26 KB
  </div>
</div>

  
  <div id="readme" class="readme blob instapaper_body">
    <article class="markdown-body entry-content" itemprop="text"><h1><a id="user-content-eukaryote-non-model-transcriptome-annotation-pipeline-entap" class="anchor" href="#eukaryote-non-model-transcriptome-annotation-pipeline-entap" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Eukaryote Non-Model Transcriptome Annotation Pipeline (enTAP)</h1>

<h2><a id="user-content-introduction" class="anchor" href="#introduction" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Introduction</h2>

<p>The software tool, enTAP, is a set of Python scripts designed to improve the accuracy and speed of functional annotations for de novo transcriptome (gene space) assemblies of non-model Eukaryotic species.</p>

<p>The scripts in this package enable the following:</p>

<ul>
<li>Integrates BLAST-style search results from up to three unique databases (and selects the most optimal annotation)</li>
<li>Identifies informative and uninformative annotations consistently</li>
<li>Identifies contaminants through optional filters for fungal, bacterial, and insect annotations.</li>
<li>Generates BLAST2GO compatible XML files to facilitate Gene Ontology term assignments</li>
<li>Integrates the results of InterProScan runs for up to two protein domain databases</li>
<li>Produces a full set of spreadsheet compatible results including summary statistics on the assembly and resulting annotations</li>
</ul>

<h2><a id="user-content-install" class="anchor" href="#install" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Install</h2>

<h3><a id="user-content-downloading-the-required-files" class="anchor" href="#downloading-the-required-files" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Downloading the required files</h3>

<p>enTAP can be downloaded with the following terminal command (Note: git is required to successfuly clone the repository):</p>

<pre><code>git clone https://github.com/SamGinzburg/WegrzynLab.git
</code></pre>

<p>The files will be placed into the "WegrzynLab/enTAP" folder.</p>

<p>The contaminant databases can be donwloaded/updated by utilizing the update.py script (usage demonstrated below)</p>

<h3><a id="user-content-dependencies" class="anchor" href="#dependencies" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Dependencies</h3>

<p>This script requires Python 2.X.   To determine the current version, please use:</p>

<pre><code>python -V
</code></pre>

<p>Python can be downloaded for free at: <a href="https://www.python.org/downloads/">https://www.python.org/downloads/</a></p>

<p>This script requires the "filter_list.txt" file to be placed in the same directory as report_generator.py in order to function correctly. This script is included in the git repository.</p>

<h3><a id="user-content-external-applications" class="anchor" href="#external-applications" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>External Applications</h3>

<p>1) Diamond is a really fast sequence-search algorithm which can compare nucleotides sequences to protein sequences for very large datasets. Alternatively blastx can also be used for small datasets.</p>

<p>Diamond download: <a href="http://github.com/bbuchfink/diamond/releases/download/v0.7.12/diamond-linux64.tar.gz">http://github.com/bbuchfink/diamond/releases/download/v0.7.12/diamond-linux64.tar.gz</a></p>

<p>Alternatively, blastx can be used for smaller datasets where diamond is slower than blastx:</p>

<p>blastx download: <a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/</a></p>

<p>2)  Blast2GO provides a full-featured GUI experience for sequence annotation.  Here, we circumvent some of the limited search options and slow network speeds by limiting its use to Gene Ontology aquisition.  Blast2GO will be run as an indepedent application during the workflow to generate a single output file.  </p>

<p>BLAST2GO download: <a href="https://www.blast2go.com/blast2go-pro/download-b2g">https://www.blast2go.com/blast2go-pro/download-b2g</a></p>

<p>3)  InterProScan is a sequence comparison tool for the identification of protein domains and wraps around several public protein domain databases.  It is used here to identify those domains and also Ontology terms associated with them.</p>

<p>InterProScan Download: <a href="https://www.ebi.ac.uk/interpro/download.html;jsessionid=B76DDDA8BCBB1AD8AFA31F3FE0E476B5">https://www.ebi.ac.uk/interpro/download.html;jsessionid=B76DDDA8BCBB1AD8AFA31F3FE0E476B5</a></p>

<h2><a id="user-content-getting-started" class="anchor" href="#getting-started" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Getting Started</h2>

<h3><a id="user-content-transcriptome-annotation-overview" class="anchor" href="#transcriptome-annotation-overview" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Transcriptome Annotation Overview</h3>

<p><a href="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/diagram.png" target="_blank"><img src="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/diagram.png" alt="Diagram of script execution" style="max-width:100%;"></a></p>

<p>The diagram above shows the order in which the seperate operations must be performed in order to generate a complete annotation report.</p>

<p>Steps:</p>

<p>1) Run your preferred search application (diamond or blastx) to obtain a correctly formatted set of search results from your database.  A total of three database runs can be integrated in the current software version.  The Blast6out tab-delimited text output option is required as is the original FASTA database used .</p>

<p>Execution example for diamond:</p>

<p>For making diamond database file:</p>

<pre><code>diamond makedb --in fastadatabase.fasta -d databasename</code></pre>

<p>For running diamond, use the following syntax:</p> 
<pre><code>diamond blastx -d /path_to_diamond_database/databasename -q query.fasta -p no_of_threads -a diamond_alignment_archive.daa -t /directory_for_temp_files --sensitive --evalue evalue_score</code></pre>

<p>For obtaining results in blast6 format, use the following command:</p>
<pre><code>diamond view -a diamond_alignment_archive.daa -o outputfilename.txt -f tab</code></pre>


<p>Execution example for blastx:</p>

<p>For making the blastx database file, use the following command:</p>

<pre><code>makeblastdb -in fastadatabase.fasta -out fastadatabase.db -dbtype prot</code></pre>

<p>For running blastx, use the following command:</p>

<pre><code>blastx -query query.fasta -db fastadatabase.db -out outputfilename.txt -outfmt 6 -num_threads no_of_threads -evalue evalue_score</code></pre>

<p>2) Run InterProScan to generate your InterProScan results file.  This can be run for up to 2 databases (specified by the appl flag).</p>

<p>Execution example for InterProScan:</p>

<p>Note: $input is binded to the configuration files for your interproscan run (refer to: <a href="https://code.google.com/p/interproscan/wiki/RC4HowToRun">https://code.google.com/p/interproscan/wiki/RC4HowToRun</a>)</p>

<pre><code>interproscan.sh -i $input -o $output -f tsv -appl pfam, Panther -goterms -iprlookup
</code></pre>

<p>3) Prepare a configuration file with the proper paths and preferences to run the report_generator.py script.  Details on the configuration file are in the section below.</p>

<pre><code>Path to query FASTA:
Source Databases (1, 2, 3): 
Sequence Search Application: (diamond)
Query Organism:
Database 1 score (1, 2, 3): 
Path to FASTA-version of database1:
Path to search results from database1:
Database 2 score (1, 2, 3):
Path to FASTA-version of database2:
Path to search results from database2:
Database 3 score (1, 2, 3):
Path to FASTA-version of database3:
Path to search results from database3:
Full-length coverage requirement (0 - 1):
Minimum Evalue (decimal form 0.00001 or scientific notation form 1e-5): 
Generate XML for Blast2GO (yes / no): 
Path to InterProScan Results:
Path to Blast2GO Results:
Contaminant Detection Options
Insects (yes / no): 
Fungi (yes / no): 
Bacteria (yes / no): 
Contaminant Database File Paths
Insects:
Fungi:
Bacteria:
</code></pre>

<p>4) Run the Report Generator script in the same directory with output files from the diamond/blastx runs and with the configuration file as the only parameter.</p>

<pre><code>cd /path/to/directory/containing/script/directory_containing_script/
python report_generator.py configuration_file.txt
</code></pre>

<p>After completing a successful execution of the report generator script, you will have a tab-delimited annotation file, a log file with summary statistics, flat files for contaminants and nohit sequences, and XML files for each database (and combined) if selected</p>

<p>5) From the combined XML file generated, you have the option to run Blast2GO to assign ontology terms:
Execution example for Blast2GO:</p>

<p>GUI Version:</p>

<p>You begin by importing your query fasta file containing your sequence information.</p>

<p><a href="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/load_sequences.png" target="_blank"><img src="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/load_sequences.png" alt="Import Sequences" style="max-width:100%;"></a></p>

<p><a href="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/load_sequences_2.png" target="_blank"><img src="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/load_sequences_2.png" alt="Import Sequences Part 2" style="max-width:100%;"></a></p>

<p>After loading your sequences you must import your Blast2GO XML file containing your sequeunce descriptions.</p>

<p><a href="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/load_xml.png" target="_blank"><img src="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/load_xml.png" alt="Import XML File" style="max-width:100%;"></a></p>

<p><a href="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/load_xml_2.png" target="_blank"><img src="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/load_xml_2.png" alt="Import XML File Part 2" style="max-width:100%;"></a></p>

<p><a href="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/load_xml_3.png" target="_blank"><img src="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/load_xml_3.png" alt="Import XML File Part 3" style="max-width:100%;"></a></p>

<p>After importing the sequences and the XML files (in that order), you must run the GO mapping step</p>

<p><a href="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/mapping.png" target="_blank"><img src="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/mapping.png" alt="Mapping Step" style="max-width:100%;"></a></p>

<p><a href="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/mapping_2.png" target="_blank"><img src="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/mapping_2.png" alt="Mapping Step" style="max-width:100%;"></a></p>

<p>After completing these steps, you have successfully mapped the GO terms, and are ready to export the Blast2GO file containing the GO terms.</p>

<p><a href="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/export_blast2go.png" target="_blank"><img src="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/export_blast2go.png" alt="Exporting the files" style="max-width:100%;"></a></p>

<p>After exporting the text file, you should get a file that looks like this:</p>

<p><a href="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/blast2go_file.png" target="_blank"><img src="https://raw.githubusercontent.com/SamGinzburg/WegrzynLab/master/enTAP/map_go_terms_instructions_images/blast2go_file.png" alt="Example File" style="max-width:100%;"></a></p>

<p>This file can then be entered back into combine_annotations.py or report_generator.py to generate a combined annotation file.</p>

<p>6) Once you have the results from InterProScan and Blast2GO, you can then run the <i>combine_annotations.py</i> script to generate your final, complete annotation file and summary statistics.</p>

<p>This script takes the Blast2GO, InterProScan, and annotation input and merges them together into one file</p>

<p>@PARAMETERS</p>

<p>--log --input, --blast2go, --interpro, --output</p>

<pre><code>python combine_annotations.py --input [annotation filename] --blast2go [blast2go filename] --interpro [interpro filename] --output [output_file_name.tsv]
</code></pre>

<p>Note this example will generate an output with the default name combined_annotation.tsv in the directory that the script is being executed in</p>

<pre><code>python combine_annotations.py --input [annotation filename] --blast2go [blast2go filename] --interpro [interpro filename]
</code></pre>

<p>The following will specify a custom output name</p>

<pre><code>python combine_annotations.py --input [annotation filename] --interpro [interpro filename] --output [output_file_name.tsv]
</code></pre>

<p>For performing merges with 2 interpro files:</p>

<pre><code>python combine_annotations.py --input [annotation filename] --interpro [interpro filename] [interpro filename two] --output [output_file_name.tsv]
</code></pre>

<p>When the script is complete, the output file will be availible in the output directory.</p>

<p>APPENDING TO AN EXISTING LOG</p>

<p>python combine_annotations.py --log [log name] --input [annotation filename] --interpro [interpro filename] [interpro filename two] --output [output_file_name.tsv]</p>

<h2><a id="user-content-configuration-file" class="anchor" href="#configuration-file" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Configuration File</h2>

<p>In the configuration file, your option is always placed after the ":". If there is no ":" on that line, then it is not an option.</p>

<p>The configuration file is line sensitive, it must always be of the form above, deleting any of the lines in the configuration file will lead to errors or unintended behavior.</p>

<pre><code>Path to query FASTA: This is the location in which you will place the path to your query fasta file

Source Databases: (1, 2, 3) Here you will enter the number of databases you wish to run your query fasta file against

Sequence Search Application: (diamond/blastx) Here you will specify the format of the search results from your databases. 

Query Organism: This setting is optional, and in the current version of the script only affects the name of the folder in which the results are placed, ex: if "walnut" was entered here the output folder would be "walnut_output_[date]"

Database 1 score: (1, 2, 3) Here you will enter the score of database 1, higher values indicate a higher preference in best hit selection

Path to FASTA-version of database1: Here you will enter the path to the fasta version of this database

Path to search results from database1: Here you will enter the path to the search results on this database

Database 2 score: (1, 2, 3) Here you will enter the score of database 2, higher values indicate a higher preference in best hit selection

Path to FASTA-version of database2: Here you will enter the path to the fasta version of this database

Path to search results from database2: Here you will enter the path to the search results on this database

Database 3 score: (1, 2, 3) Here you will enter the score of database 3, higher values indicate a higher preference in best hit selection

Path to FASTA-version of database3: Here you will enter the path to the fasta version of this database

Path to search results from database3: Here you will enter the path to the search results on this database

Full-length coverage requirement: (0 - 1, not including 0) Here you will enter the full length coverage requirement (recommended starting value = 0.7)

Minimum Evalue: (decimal form 0.00001 or scientific notation form 1e-5) Here you will enter the minimum e-value to be used in best hit selection (recommended starting value = 1e-5)

Generate XML for Blast2GO: (yes / no)

Path to InterProScan Results: Here you will enter the path to your InterProScan results (if you have them), these results can also be appended later via the combine_annotations.py script.

Path to Blast2GO Results: Here you will enter the path to your Blast2GO results (if you have them), these results can also be appended later via the combine_annotations.py script.

Contaminant Detection Options // Select yes to include these hits when running the script, and no to not include them.

Insects: (yes / no)
Fungi: (yes / no)
Bacteria: (yes / no)

Contaminant Database File Paths // These databases can be downloaded using the update.py script

Insects: Path to the insects contaminant database
Fungi: Path to the fungi contaminant database
Bacteria: Path to the bacteria contaminant database
</code></pre>

<h4><a id="user-content-query-fasta" class="anchor" href="#query-fasta" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Query Fasta</h4>

<p>At least one database run is necessary and the query FASTA represents the transcriptome assembly and this should be a multi-FASTA text file:</p>

<pre><code>&gt;Contig_PtHGO
ATTAAATCCAAACACTATTACTGTGACATTAACAAGGAGAGCAACGGCCTCTGTCTGTATTTGTGTTTGGCAGACGATATTTATGGATTCGCAGCACGTCAATATTTCATGAAAAAAATATCGCCCTAATATAAGATAATCATAACTTTGTTTTTCCTTTGAACTTTACTTCAACCGAAGATCTCGACTGCTGAATTCTGAATTCTGCAATTTGAATGCGCGGGGTTAAAAATGGAGAAATTGGCGAAGCTAAACATTTCAGAAGAGGACTTGGAATTCGAATTTAGCCATCTGGAATACCAGAGCGGATTTGGCAATAATTTTTCCTCGGAAGCAAAACCTGGTGCACTGCCGGTGGGGCAGAATAATCCCTTGAAGTGTCCTTATGGCCTCTACGCCGAGCAAATCTCAGGGACGGCATTTACAGTCCCCAGGAAGCTCAACCAGAGAAGTTGGCTGTACCGTATCAAGCCATCAGTAACACATGAACCTTTTCATCCTCGGGTTCCATCACATGATTATCTGATAAGTGAGTTCAATCAGTCATCAAGCTCAGCAACACCAACGCAATTGCGTTGGAATCCTTTCAACATTCCAGATACGAAAACAAATTTCATTGATGGACTGTTCACTATATGTGGAGCTGGTAGTTCATTCCTTCGGCATGGCTTTGCAGTGCACATGTATGTCGCAAATGCATCTATGGAAGGCTGTGCATTTGCTAATGCTGATGGGGACTTTCTGATAGTTCCACAACAAGGAAGATTATGGATAACGACAGAGTTAGGAAGATTGCAGGTTTCTCCTGGTGAAATAGTTGTGTTGCAACAAGGTTTCCGGTATTCCATTGATTTACCAGATGGTCCTTCTCGAGGGTACGTTGTAGAGGTTTTCAGTGGCCATTTCCAGCTTCCCGATCTTGGTCCAATCGGTGCAAATGGCCTGGCCTCACCGCCTGACTTTCTAACACCTACAGCATGGTTTGATGACAAAGCCTATCCTGGTTACACGATTGTTCATAAGTTTGGAGGTTCATTGTTTACAGCAAAACAAAATTTCTCCCCCTTCAATGTGGTTGCTTGGCATGGCAATTATGCACCTTACAAGTATGATTTGAAGAAGTTCTGTCCTTTCAATACTGTTTTATTTGATCATGGTGACCCATCAATAAACACAGTTCTGACAGTTCCATCAGAAAAACCAGGAGTAGCTGTGGTTGACTTTGTCATTTTTCCACCTCGCTGGCTGGTAGCAGAGCATACCTTCCGTCCTCCATATTACCACCGTAACTGCATGAGTGAATTTATGGGCCTGATTTATGGCAGCTATGAGGCTAAAGCAGATGGATTTCAACCAGGAGGGGCAAGCTTGCACAGCTGCATGACCCCACATGGGCCTGATACTACCACGTTTGAGAAAACTATAGCTGAAGAGGGCGATGCAAAACCTGCCAAGATAAAAGATACCATGGCTTTCATGTTTGAGTCATCATTGATTCCCAGAATCACACCATGGGTTCTCAAATCTCCCCATTTGGATAACGATTATTACAAGTGCTGGACAGGTCTCAAGTCCCATTTTCACCATGAACATCTTCCTGAAAACGGGGAATTACATAGTTAGTAAATCTGTTTCTGTAAATATGGAGAGATATGTACT

&gt;Contig_PtHPPD
ATTTAGTGCAAGAAACAATGGGTAAGGATGGGAGTCAGAGCTTTAAACTGGTGGGATACAAGAACTTTGTTCGCCACAATCCCATGAGCGATCGCTTCCAAGTGAAGAAATTCCACCACATTGAGTTCTGGTGCGGCGATGCCACCAATACCCACAAGCGCTTCTCATGGGGTTTGGGCATGCAGGTGGTGGCCAAGTCCGATCAGAGCACCGGAAATCAGACATATTGCAGCTTCGCCATACAGTCCAACGATCTCACCTTAGTCTTCACGGCGCCTTATTCTTCGAAAATTGACCAGTCCACCACCAGAATGCCCCACCCAGCTTTCGATTCCGATGAGGCGCGGAGTTTCTTTGTCAAGCATGGGTTAGCAGTTCGAGCCGTGGGAGTGCTTGTCGAAGATGCTCGAGACGCTTTCCAGAATAGCGTGGAAAACGGTGCCGTGGCTGTGCTGGAGCCATGCGATCTTGTCGATGAGGCCACGGGCGGGAAGGTCAGTATGTCGGAGGTTAAGCTTTACGGAGATGTGGTTCTTCGTTTTGTAAGCCAGGACGGTTACAAGGGCCCCTTTTTGCCCAATTACGAGCCCGTTCAGTCGGTTCAATTGTCCTATGGAATCATTCGGGTCGACCACGCTGTGGGCAATGTGGAGAAACTGGAGGAAGCCGTCCAATaTGTGGCCAAGTTTAGCGGGTTTCACCGTTTTGCCGAGTTTACGGCAGAAGACGTCGGGACTGGAGAGAGTGGCCTGAATTCGATGGTGTTGGCGAGCAATAATGAGATGGTTCTATTGCCCATGAACGAGCCTGTGTTTGGGACCAAAAGGAAGTCACAGATACAGACCTATCTGGAGCACAATGAGGGCCCCGGATTGCAGCATTTGGCTCTTATTTGCTCTGATATATTCTCCACGCTTAAGGAGATGCGCATGCGCAGTGAAATTGGAGGGTTTGAATTCATGCCCAAGCCTCCTCCTACTTATTATAAGAATCTGAGGAATAGGGTTGCAGATGTCTTGACTGAGAAGCAGATGGAAGAGTGCGACGAGTTGGGGATTTTGGTCGATAGGGATGAtCAAGGAGTCCTCTTACAGATTTTCACAAAGCCCATTGGCGACAGGCCCACTATTTTTATTGAGATAATTCAACGTGTAGGCTGTGTGCTAACTGATGAGAATGGGAAAGCCTATCAAAAGGGTGGATGTGGAGGATTTGGCAAAGGCAATTTTTCAGAACTCTTTAAATCCATCGAGGAATACGAGAAAACACTTGAAGGAAGAGTCAATTCCTAGGCCCTTCGAAGTTCGTAGTGATCTCCTAGAGTAGATTCATCTTGGAATTAGACTGATAGAGATTTATTATATACGTGAGAGTGGCATAATGATAAAATACATGTACTACGGTTGCTTGTGTTAATGGTTTCTAGATGTCGTTTAGTGAAGATGCCTTGTAAAGAAGCAAGGTATATATGTTATACAAATATACTCTGTTTCTGGTTTGGTTCTAGGCTGCATGCCCTTAAGAGGAATTTAACTGTTATAATTTAAATATATTATGTTCAGATAGGCCCACGGCTTATCATTTTATAATTTAAATATATTCTGTTCAGATATGTCCACGGCTGGTTAAGTTCACTTGTCAAGTAACTCCATTTTTATTCACGCTGGAGCGTAATTTTGGATCTAGGGATATAGAATCAGGATGGATGCTCTCTTTAAAGCTTATAACATATTTATGGTTCGT

</code></pre>

<h4><a id="user-content-source-databases" class="anchor" href="#source-databases" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Source Databases</h4>

<p>This option contains the number of databases to be parsed by the script, and can be a number between 0 and 3.</p>

<h4><a id="user-content-sequence-search-application" class="anchor" href="#sequence-search-application" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Sequence Search Application</h4>

<p>This script currently supports 2 different sequence search applications, namely diamond and blastx.</p>

<p>When running these applications, it is important to use the blast6 format.</p>

<p>The correct format looks like this:</p>

<pre><code>Contig_PtHGO    gi|460412999|ref|XP_004251883.1|        78.7    432     92      0       286     1581    12      443     4.3e-212        744.2
</code></pre>

<h4><a id="user-content-paths-to-the-databases" class="anchor" href="#paths-to-the-databases" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Paths to the Databases</h4>

<p>The current version of the script does not support indexed versions of databases so the flat file (multi-FASTA version) of the database must be supplied for each database queried.</p>

<h4><a id="user-content-minimum-e-value-and-minimum-coverage-requirements" class="anchor" href="#minimum-e-value-and-minimum-coverage-requirements" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Minimum E-Value and Minimum coverage requirements</h4>

<p>The minimum E-Value can be input in either scientific notation (1e-5) or as a decimal (0.00001).</p>

<p>The minimum coverage requirement is a value between 0 and 1 where 70% coverage represented as 0.7 is the default value.</p>

<h4><a id="user-content-interproscan-and-blast2go-paths" class="anchor" href="#interproscan-and-blast2go-paths" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>InterProScan and Blast2GO paths</h4>

<p>For these two options you must enter the path to your InterProScan and Blast2GO output file (Sequence Table format).</p>

<p>The InterProScan file format is as follows:</p>

<pre><code>c10004_g1_i1    3F8F69B8D17CEA7F        185     HMMPfam PF13976 gag_pre-integrs 37      108     1.0E-10 T       06-Nov-2014     IPR025724       GAG-pre-integrase domain
</code></pre>

<p>The Blast2GO Sequence Table format is as follows:</p>

<pre><code>c58101_g1_i1|m.1        uninformative   390     2       7.2E-10 0.0%    3       F:GO:0003674; P:GO:0008150; C:GO:0005575
</code></pre>

<h4><a id="user-content-contaminant-db-selection" class="anchor" href="#contaminant-db-selection" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Contaminant DB selection</h4>

<p>This script currently has three optional flat file databases containing genus and species information of organisms that could be filtered as potential contaminants in Eukaryote transcriptome studies:</p>

<ul>
<li>Insects</li>
<li>Bacteria</li>
<li>Fungi</li>
</ul>

<p>In in configuration file, each of these can be toggled by writing either "y" or "yes" to include them when searching for contaminants, or write "no" or leave it blank to ignore these filters.</p>

<h4><a id="user-content-downloading-updated-contaminant-databases-using-updatepy" class="anchor" href="#downloading-updated-contaminant-databases-using-updatepy" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Downloading updated contaminant databases using update.py</h4>

<p>The update.py script will automatically download all of the contaminant files required to run report_generator.py</p>

<p>The files will be deposited into a folder titled "contaminant_databases".</p>

<p>correct update.py usage:</p>

<pre><code>python update.py
</code></pre>

<p>Note: An internet connection is required to run update.py.</p>

<h2><a id="user-content-output" class="anchor" href="#output" aria-hidden="true"><svg aria-hidden="true" class="octicon octicon-link" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M4 9h1v1h-1c-1.5 0-3-1.69-3-3.5s1.55-3.5 3-3.5h4c1.45 0 3 1.69 3 3.5 0 1.41-0.91 2.72-2 3.25v-1.16c0.58-0.45 1-1.27 1-2.09 0-1.28-1.02-2.5-2-2.5H4c-0.98 0-2 1.22-2 2.5s1 2.5 2 2.5z m9-3h-1v1h1c1 0 2 1.22 2 2.5s-1.02 2.5-2 2.5H9c-0.98 0-2-1.22-2-2.5 0-0.83 0.42-1.64 1-2.09v-1.16c-1.09 0.53-2 1.84-2 3.25 0 1.81 1.55 3.5 3 3.5h4c1.45 0 3-1.69 3-3.5s-1.5-3.5-3-3.5z"></path></svg></a>Output</h2>

<p>The script outputs several files.</p>

<p>There will be a text file of the name: log_[date of execution goes here].txt </p>

<p>This file will contain the log with all of the statistical data generated in respect to each database, and the collection of databases as a whole.
The file will be of the form:</p>

<pre><code>Weighted Summary Statistics on the Transcriptome Assembly Input (Query)
Total Number of Query Sequences:        29785
Median Query Length:    825
Average Query Length:   1048.32
Shortest Query Length:  297
Longest Query Length:   15345
N50 Statistic:  1344
Number of queries with an informative hit:      16183
Number of queries with an uninformative hit:    9404
Number of queries with no hit:  4140
Number of contaminants:         58
The top 10 hits by species:
uninformative:  9404
PREDICTED: putative ribonuclease H protein At1g65750-like:      94
ATP binding protein, putative:  79
DNA binding protein, putative:  64
transcription factor, putative: 60
membrane protein:       59
PREDICTED: TMV resistance protein N-like:       46
protein binding protein, putative:      45
zinc finger protein, putative:  41
transcriptional regulator:      35
The top 10 contaminants:
Escherichia coli        28
Laccaria bicolor        4
Zymoseptoria tritici    3
Coprinopsis cinerea     3
Magnaporthe oryzae      2
Ustilago maydis 2
Pyrenophora teres       2
Cryptococcus gattii     2
Pyrenophora tritici-repentis    2
Mycobacterium kansasii  1

Summary Statistics on the Transcriptome Assembly Input (Query)
Total Number of Query Sequences:        29785
Median Query Length:    825
Average Query Length:   1048.32
Shortest Query Length:  297
Longest Query Length:   15345
N50 Statistic:  1344
Number of queries with an informative hit:      16183
Number of queries with an uninformative hit:    9404
Number of queries with no hit:  4140
Number of contaminants:         58
The top 10 hits by species:
uninformative:  9404
PREDICTED: putative ribonuclease H protein At1g65750-like:      94
ATP binding protein, putative:  79
DNA binding protein, putative:  64
transcription factor, putative: 60
membrane protein:       59
PREDICTED: TMV resistance protein N-like:       46
protein binding protein, putative:      45
zinc finger protein, putative:  41
transcriptional regulator:      35
The top 10 contaminants:
Escherichia coli        28
Laccaria bicolor        4
Zymoseptoria tritici    3
Coprinopsis cinerea     3
Magnaporthe oryzae      2
Ustilago maydis 2
Pyrenophora teres       2
Cryptococcus gattii     2
Pyrenophora tritici-repentis    2
Mycobacterium kansasii  1

DB:     /path/to/file/combined_genes_blast6out_ppf.out
Number of queries with an informative hit:      13723
Number of queries with an uninformative hit:    10857
Number of contaminants: 0
Number of queries with no hit:  5205
The top 10 hits by species:
uninformative:  10857
PREDICTED: putative ribonuclease H protein At1g65750-like:      85
ATP binding protein, putative:  62
DNA binding protein, putative:  48
PREDICTED: TMV resistance protein N-like:       45
transcription factor, putative: 40
protein binding protein, putative:      38
pentatricopeptide repeat-containing protein, putative:  29
PREDICTED: UPF0481 protein At3g47200-like:      29
zinc finger family protein:     28
The top 10 contaminants by species:
No contaminants present

DB:     /path/to/file/combined_genes_blast6out_rsp.out
Number of queries with an informative hit:      16103
Number of queries with an uninformative hit:    9196
Number of contaminants: 58
Number of queries with no hit:  4428
The top 10 hits by species:
uninformative:  9196
PREDICTED: putative ribonuclease H protein At1g65750-like:      85
ATP binding protein, putative:  78
DNA binding protein, putative:  64
transcription factor, putative: 60
membrane protein:       59
protein binding protein, putative:      45
PREDICTED: TMV resistance protein N-like:       44
zinc finger protein, putative:  41
transcriptional regulator:      35
The top 10 contaminants by species:
Escherichia coli:       28
Laccaria bicolor:       4
Zymoseptoria tritici:   3
Coprinopsis cinerea:    3
Magnaporthe oryzae:     2
Ustilago maydis:        2
Pyrenophora teres:      2
Cryptococcus gattii:    2
Pyrenophora tritici-repentis:   2
Mycobacterium kansasii:         1

Interpro File:  /path/to/file/interpro.raw
Blast2GO File:  /path/to/file/blast2go_walnut_transcriptome.txt
Number of sequences with Domain Identification:         25587
Number of sequences without Domain Identification:      0
Blast2GO Gene Ontology Stats
Number of Components:   4552
Number of Functions:    326
Number of Processes:    4986
Number of Transcripts with at least 1 Component:        4552
Number of Transcripts with at least 1 Function:         4986
Number of Transcripts with at least 1 Process:  326
Interpro Gene Ontology Stats (Totals)
Component:      2859
Function:       8690
Process:        13303
Number of Transcripts with at least 1 Component:        2421
Number of Transcripts with at least 1 Function:         7622
Number of Transcripts with at least 1 Process:  6826
</code></pre>

<p>There will be a text file of the name: nohits_[date of execution goes here].txt
The file will be of the form:</p>

<pre><code>c54972_g2_i2|m.50780
c54259_g1_i5|m.2219
c37958_g2_i1|m.6596
c55455_g2_i1|m.35545
c51509_g3_i1|m.34497
c54913_g5_i1|m.11269
c29778_g1_i1|m.6050
c970_g1_i1|m.12443
c53384_g7_i1|m.30991
</code></pre>

<p>This file will contain the log of all of the nohits found while searching for matches to the search queries in the search results from the databases.</p>

<p>There will be a series of XML files of the name: blastxml_[db number OR combined_db]_[date of execution goes here].xml </p>

<p>These are the file(s) that can be imported back into Blast2GO as an XML file.</p>

<p>There will be a file of the name: uninformative_[date of execution goes here].txt </p>

<p>This file will contain the log of all the uninformative hits.</p>

<p>There will be a file of the name: default_output_annotation_2015-03-22 13:07:41.898127.tsv</p>

<p>This file will contain all of the search queries from the query fasta that had a hit, along with their best hit and description.</p>

<p>This file will be of the form:</p>

<pre><code>Query Subject_id Identity(%) Alignment_length Mismatches Number of gap opens Query_start Query_end Subject_start Subject_end E-value Bit_score Origin Database Subject Description Species
c48227_g1_i1|m.33460 gi|255540671|ref|XP_002511400.1| 66.2 542 168 7 1 1605 1 534 1.4e-198 699.5 /group/nealedata3/Walnut_Transcriptome/large_set/combined/combined_genes_blast6out_rsp.out uninformative Ricinus communis
</code></pre>

<p>There will be a file of the name: combined_annotation_[date of execution goes here].tsv</p>

<p>This file will be of the form:</p>

<pre><code>Query Subject_id Identity(%) Alignment_length Mismatches Number of gap opens Query_start Query_end Subject_start Subject_end E-value Bit_score Origin Database Subject Description Species Signature Description InterPro accession number InterPro description blast2go_process blast2go_function blast2go_component
c48227_g1_i1|m.33460 gi|255540671|ref|XP_002511400.1| 66.2 542 168 7 1 1605 1 534 1.4e-198 699.5 /group/nealedata3/Walnut_Transcriptome/large_set/combined/combined_genes_blast6out_rsp.out uninformative Ricinus communis Pex14_N IPR006785 Peroxisome membrane anchor protein Pex14p, N-terminal
c55482_g1_i1|m.33976 gi|449513131|ref|XP_004164240.1| 50.1 814 367 16 28 2406 8 803 2.5e-207 729.2 /group/nealedata3/Walnut_Transcriptome/large_set/combined/combined_genes_blast6out_rsp.out PREDICTED: probable receptor-like protein kinase At2g23200-like Cucumis sativus Pkinase_Tyr, Malectin_like IPR001245, IPR024788 Serine-threonine/tyrosine-protein kinase catalytic domain, Malectin-like carbohydrate-binding domain Molecular Function:protein kinase activity (GO:0004672), Biological Process:protein phosphorylation (GO:0006468)  F:protein serine/threonine kinase activity
</code></pre>

<p>This file will contain all of the search queries from the query fasta that had a hit, their respective best hits, descriptions, and InterProScan and Blast2GO results.</p>
</article>
  </div>

</div>

<button type="button" data-facebox="#jump-to-line" data-facebox-class="linejump" data-hotkey="l" class="hidden">Jump to Line</button>
<div id="jump-to-line" style="display:none">
  <!-- </textarea> --><!-- '"` --><form accept-charset="UTF-8" action="" class="js-jump-to-line-form" method="get"><div style="margin:0;padding:0;display:inline"><input name="utf8" type="hidden" value="&#x2713;" /></div>
    <input class="form-control linejump-input js-jump-to-line-field" type="text" placeholder="Jump to line&hellip;" aria-label="Jump to line" autofocus>
    <button type="submit" class="btn">Go</button>
</form></div>

  </div>
  <div class="modal-backdrop"></div>
</div>


    </div>
  </div>

    </div>

        <div class="container site-footer-container">
  <div class="site-footer" role="contentinfo">
    <ul class="site-footer-links right">
        <li><a href="https://status.github.com/" data-ga-click="Footer, go to status, text:status">Status</a></li>
      <li><a href="https://developer.github.com" data-ga-click="Footer, go to api, text:api">API</a></li>
      <li><a href="https://training.github.com" data-ga-click="Footer, go to training, text:training">Training</a></li>
      <li><a href="https://shop.github.com" data-ga-click="Footer, go to shop, text:shop">Shop</a></li>
        <li><a href="https://github.com/blog" data-ga-click="Footer, go to blog, text:blog">Blog</a></li>
        <li><a href="https://github.com/about" data-ga-click="Footer, go to about, text:about">About</a></li>

    </ul>

    <a href="https://github.com" aria-label="Homepage" class="site-footer-mark">
      <svg aria-hidden="true" class="octicon octicon-mark-github" height="24" title="GitHub " version="1.1" viewBox="0 0 16 16" width="24"><path d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59 0.4 0.07 0.55-0.17 0.55-0.38 0-0.19-0.01-0.82-0.01-1.49-2.01 0.37-2.53-0.49-2.69-0.94-0.09-0.23-0.48-0.94-0.82-1.13-0.28-0.15-0.68-0.52-0.01-0.53 0.63-0.01 1.08 0.58 1.23 0.82 0.72 1.21 1.87 0.87 2.33 0.66 0.07-0.52 0.28-0.87 0.51-1.07-1.78-0.2-3.64-0.89-3.64-3.95 0-0.87 0.31-1.59 0.82-2.15-0.08-0.2-0.36-1.02 0.08-2.12 0 0 0.67-0.21 2.2 0.82 0.64-0.18 1.32-0.27 2-0.27 0.68 0 1.36 0.09 2 0.27 1.53-1.04 2.2-0.82 2.2-0.82 0.44 1.1 0.16 1.92 0.08 2.12 0.51 0.56 0.82 1.27 0.82 2.15 0 3.07-1.87 3.75-3.65 3.95 0.29 0.25 0.54 0.73 0.54 1.48 0 1.07-0.01 1.93-0.01 2.2 0 0.21 0.15 0.46 0.55 0.38C13.71 14.53 16 11.53 16 8 16 3.58 12.42 0 8 0z"></path></svg>
</a>
    <ul class="site-footer-links">
      <li>&copy; 2016 <span title="0.14564s from github-fe156-cp1-prd.iad.github.net">GitHub</span>, Inc.</li>
        <li><a href="https://github.com/site/terms" data-ga-click="Footer, go to terms, text:terms">Terms</a></li>
        <li><a href="https://github.com/site/privacy" data-ga-click="Footer, go to privacy, text:privacy">Privacy</a></li>
        <li><a href="https://github.com/security" data-ga-click="Footer, go to security, text:security">Security</a></li>
        <li><a href="https://github.com/contact" data-ga-click="Footer, go to contact, text:contact">Contact</a></li>
        <li><a href="https://help.github.com" data-ga-click="Footer, go to help, text:help">Help</a></li>
    </ul>
  </div>
</div>



    
    

    <div id="ajax-error-message" class="ajax-error-message flash flash-error">
      <svg aria-hidden="true" class="octicon octicon-alert" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M15.72 12.5l-6.85-11.98C8.69 0.21 8.36 0.02 8 0.02s-0.69 0.19-0.87 0.5l-6.85 11.98c-0.18 0.31-0.18 0.69 0 1C0.47 13.81 0.8 14 1.15 14h13.7c0.36 0 0.69-0.19 0.86-0.5S15.89 12.81 15.72 12.5zM9 12H7V10h2V12zM9 9H7V5h2V9z"></path></svg>
      <button type="button" class="flash-close js-flash-close js-ajax-error-dismiss" aria-label="Dismiss error">
        <svg aria-hidden="true" class="octicon octicon-x" height="16" version="1.1" viewBox="0 0 12 16" width="12"><path d="M7.48 8l3.75 3.75-1.48 1.48-3.75-3.75-3.75 3.75-1.48-1.48 3.75-3.75L0.77 4.25l1.48-1.48 3.75 3.75 3.75-3.75 1.48 1.48-3.75 3.75z"></path></svg>
      </button>
      Something went wrong with that request. Please try again.
    </div>


      
      <script crossorigin="anonymous" src="https://assets-cdn.github.com/assets/frameworks-4e4bd0cf3c442c849b926e75ec4e55233bc74e65771dc70d93aa133b32705488.js"></script>
      <script async="async" crossorigin="anonymous" src="https://assets-cdn.github.com/assets/github-97fe248a925294b2a437a86d80688ff0684f57be07085dda8e893ff5d9a5338b.js"></script>
      
      
      
      
    <div class="js-stale-session-flash stale-session-flash flash flash-warn flash-banner hidden">
      <svg aria-hidden="true" class="octicon octicon-alert" height="16" version="1.1" viewBox="0 0 16 16" width="16"><path d="M15.72 12.5l-6.85-11.98C8.69 0.21 8.36 0.02 8 0.02s-0.69 0.19-0.87 0.5l-6.85 11.98c-0.18 0.31-0.18 0.69 0 1C0.47 13.81 0.8 14 1.15 14h13.7c0.36 0 0.69-0.19 0.86-0.5S15.89 12.81 15.72 12.5zM9 12H7V10h2V12zM9 9H7V5h2V9z"></path></svg>
      <span class="signed-in-tab-flash">You signed in with another tab or window. <a href="">Reload</a> to refresh your session.</span>
      <span class="signed-out-tab-flash">You signed out in another tab or window. <a href="">Reload</a> to refresh your session.</span>
    </div>
    <div class="facebox" id="facebox" style="display:none;">
  <div class="facebox-popup">
    <div class="facebox-content" role="dialog" aria-labelledby="facebox-header" aria-describedby="facebox-description">
    </div>
    <button type="button" class="facebox-close js-facebox-close" aria-label="Close modal">
      <svg aria-hidden="true" class="octicon octicon-x" height="16" version="1.1" viewBox="0 0 12 16" width="12"><path d="M7.48 8l3.75 3.75-1.48 1.48-3.75-3.75-3.75 3.75-1.48-1.48 3.75-3.75L0.77 4.25l1.48-1.48 3.75 3.75 3.75-3.75 1.48 1.48-3.75 3.75z"></path></svg>
    </button>
  </div>
</div>

  </body>
</html>

